### PR TITLE
RHMAP-11589 - create delete api to delete collections

### PR DESCRIPTION
**Pull request by rachael-oregan**
# Motivation
Create an API to delete selected collection(s)

# Result
A delete endpoint that takes an array of collection names as parameters, checks if the collections requested for deletion exist in the database and then deletes them.

# Jira
https://issues.jboss.org/browse/RHMAP-11589

### CODE REVIEW

File: src/endpoints/http/collections/delete.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * 
+ * @param  {Object} router - The express router.
+ */
+export function deleteCollections(router) {
+  router.post('/collections', (req, res, next) => {
+    const reqCollections = req.body.names;
+    if (!reqCollections) {
+      return res.status(400).send(message('names(s) of collection(s) is required'));
+    }
+    const allCollectionNames = req.db.getCollectionNames();
```
Comment (wei-lee) [inline src/endpoints/http/collections/delete.js]: I would consider the code from here is the business logic. I would prefer us to move the business logic out of the controller, even if sometimes the business logic is relatively simple. I think there are a few benefits:

1. loose coupling between controllers and the business logic. Make it easier to move the business logic if we need to.
2. easier to test the business logic.
3. keep things consistent. Make it looks like the code is written by one developer.

See https://github.com/feedhenry/fh-dataman/blob/master/src/endpoints/http/collections/index.js#L4 for example.

File: src/endpoints/http/collections/delete.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    if (!allCollectionNames) {
+      return res.status(400).send(message('No collections exists in database'));
+    }
+    _.filter(reqCollections, function(name) {
```
Comment (wei-lee) [inline src/endpoints/http/collections/delete.js]: I think we should first check all the collection names are actually valid. If there is any collection name do not exist, return the error.

File: src/endpoints/http/collections/delete.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      if (_.includes(allCollectionNames, name)) {
+        const collectionToDelete = req.db.getCollection(name);
+        req.log.debug({name}, 'deleting collection');
+        req.db.collectionToDelete.drop();
```
Comment (wei-lee) [inline src/endpoints/http/collections/delete.js]: perhaps consider using Promise.all or async.series here to wait for the collection drop action to complete. At the moment, the user will not know even if the collection failed to drop

File: src/endpoints/http/collections/delete_test.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import bodyParser from 'body-parser';
```
Comment (wei-lee) [inline src/endpoints/http/collections/delete_test.js]: if we separate out the business logic (delete collections) out of the controller, this test can be easily written without the need of supertest

**Review by wei-lee — CHANGES_REQUESTED** [review]
we should also add an integration test for the delete action

File: src/endpoints/http/collections/delete.js (hunk 2 above)
Comment (rachael-oregan) [reply src/endpoints/http/collections/delete.js]: The way I'm checking if the collection names are valid is through the `includes()` function. The `filter()` function iterates over the array `reqCollections` and if an element exists in the `allCollectionNames` array then I get the collection by `req.db.getCollection(name)`. I'll add in that an error is returned if it doesn't exist

File: src/endpoints/http/collections/delete.js (hunk 1 above)
Comment (rachael-oregan) [reply src/endpoints/http/collections/delete.js]: That's grand I'll move the business logic. Just wondering, why is the business logic in the create controller?

Comment (wei-lee) [reply src/endpoints/http/collections/delete.js]: @rachael-oregan  which create controller?

Comment (rachael-oregan) [reply src/endpoints/http/collections/delete.js]: @wei-lee https://github.com/feedhenry/fh-dataman/blob/master/src/endpoints/http/collections/create.js#L11

Comment (wei-lee) [reply src/endpoints/http/collections/delete.js]: @rachael-oregan right. Yeah, we might need to refactor the creation to keep things consistent.

**Pull request closed without merging**